### DISCUSSION TITLE
RequestValidationError can't json serialize some inputs

### DISCUSSION THREAD

**Discussion by bertomaniac** [fastapi/fastapi#11923] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from io import BytesIO
from json import dumps
from fastapi import FastAPI
from fastapi.testclient import TestClient


def test_request_validation_error_with_binary():
    app = FastAPI()

    @app.post("/request-validation")
    def route_with_json_data(
        data: dict,
    ):
        return data

    client = TestClient(app)

    post_data = {"data": "some_field"}

    # simulate binary image data, the \xff throws a decode error
    file_binary = BytesIO(b"some initial binary data: \xff\x01")
    
    # here we're calling a json endpoint route with non-json data
    # I'd expect this to be just a standard 422 response
    response = client.post(
        "/request-validation",
        files={"filestream": ("filename", file_binary)},
        data={"data": dumps(post_data)},
    )

    # this _should_ come back as a 422 but instead throws a
    # UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 283: invalid start byte
    assert response.status_code == 422
```

### Description

The crux of the issue is that if body data is sent to a json endpoint, the `request_validation_exception_handler` attempts to serialize the body data back to the user and, in some cases, improperly serializes it and throws an error.

This _used_ to work in `fastapi` with `pydantic v1` but it seems the new `ValidationError` structure in pydantic v2 has changed how this works. Specifically, when `ValidationError.errors()` is called, it includes the `input` as part of the error structure. 

The fastapi code does properly generate the validation errors, but when attempting to json serialize the error back to the user, it chokes for certain data types.

I feel like the default `request_validation_exception_handler` included in `fastapi` should handle this or perhaps there should be an option to provide overrides to the `ValidationError.errors()` call (similar to https://github.com/fastapi/fastapi/discussions/10934)

### Operating System

macOS

### FastAPI Version

0.111.0

### Pydantic Version

2.8.2

### Python Version

3.12.4

**Comment by bertomaniac** [comment, marked as answer]
I ended up fixing this problem by adding an exception handler for the `RequestValidationError` to strip out the `input` field before json serializing the error:

```
from fastapi.exceptions import RequestValidationError
from pydantic import ValidationError

# adapted from the Pydantic docs:
# https://docs.pydantic.dev/latest/errors/errors/#custom-errors
def convert_validation_errors(validation_error: ValidationError | RequestValidationError) -> list[dict[str, Any]]:
    converted_errors = []
    for error in validation_error.errors():
        converted_error = {
            "type": error["type"],
            "loc": loc_to_dot_sep(error["loc"]),
            "msg": error["msg"],
        }
        converted_errors.append(converted_error)
    return converted_errors


def loc_to_dot_sep(loc: tuple[str | int, ...]) -> str:
    path = ""
    for i, x in enumerate(loc):
        if isinstance(x, str):
            if i > 0:
                path += "."
            path += x
        elif isinstance(x, int):
            path += f"[{x}]"
        else:
            raise TypeError("Unexpected type")
    return path


async def request_validation_exception_handler(request: Request, exc: RequestValidationError) -> JSONResponse:
    errors = convert_validation_errors(exc)
    return JSONResponse(
        status_code=HTTPStatus.UNPROCESSABLE_ENTITY,
        content={"detail": jsonable_encoder(errors)},
    )
```

This seems to be the simplest fix so leaving here in case others run into this issue.

**Comment by RichieB2B** [comment]
We ran into the same issue, thanks a lot for providing a fix! The above code is missing some imports. What worked for me was:
```
from http import HTTPStatus
from typing import Any

from fastapi import Request
from fastapi.encoders import jsonable_encoder
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from pydantic import ValidationError
```

Then I was able to use the new `request_validation_exception_handle` by adding it to my FastAPI app:
```
app = FastAPI()
app.add_exception_handler(RequestValidationError, request_validation_exception_handler)
```

**Comment by RichieB2B** [comment]
It would be much nicer if `input` is only omitted when `jsonable_encoder` fails. The full solution I am now using:
```
from http import HTTPStatus
from typing import Any

from fastapi import Request
from fastapi.encoders import jsonable_encoder
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from pydantic import ValidationError


# adapted from the Pydantic docs:
# https://docs.pydantic.dev/latest/errors/errors/#custom-errors
def convert_validation_errors(validation_error: ValidationError | RequestValidationError) -> list[dict[str, Any]]:
    converted_errors = []
    for error in validation_error.errors():
        try:
            jsonable_encoder([error])
            converted_errors.append(error)
        except UnicodeDecodeError:
            converted_error = {
                "type": error["type"],
                "loc": loc_to_dot_sep(error["loc"]),
                "msg": error["msg"],
            }
            converted_errors.append(converted_error)
    return converted_errors


def loc_to_dot_sep(loc: tuple[str | int, ...]) -> str:
    path = ""
    for i, x in enumerate(loc):
        if isinstance(x, str):
            if i > 0:
                path += "."
            path += x
        elif isinstance(x, int):
            path += f"[{x}]"
        else:
            raise TypeError("Unexpected type")
    return path


async def request_validation_exception_handler(request: Request, exc: RequestValidationError) -> JSONResponse:
    errors = convert_validation_errors(exc)
    return JSONResponse(
        status_code=HTTPStatus.UNPROCESSABLE_ENTITY,
        content={"detail": jsonable_encoder(errors)},
    )
```
